### PR TITLE
DDCNL-7863: Replaced SafeRedirectUrl by RedirectUrl with policy

### PR DESCRIPTION
**Pull request by askar-makkunnath**

### CODE REVIEW

File: app/views/interstitial/DisplayAddressInterstitialView.scala.html [hunk 1]
```diff
@@ -33,6 +33,6 @@
 
     @h1("label.profile")
     @p(Text(messages("label.you_can_see_this_part_of_your_account_if_you_complete")))
-    @button(msg = "global.label.continue_button", href=Some(controllers.routes.ApplicationController.uplift(Some(SafeRedirectUrl(continueUrl))).url))
+    @button(msg = "global.label.continue_button", href=Some(controllers.routes.ApplicationController.uplift(Some(RedirectUrl(continueUrl))).url))
```
Comment (vfryer12) [inline app/views/interstitial/DisplayAddressInterstitialView.scala.html:36]: The import needs to be updated to @import uk.gov.hmrc.play.bootstrap.binders.RedirectUrl

File: app/views/interstitial/ViewChildBenefitsSummarySingleAccountInterstitialView.scala.html [hunk 2]
```diff
@@ -91,6 +91,6 @@
             </ul>
         } else {
             @p(Text(messages("label.you_can_see_this_part_of_your_account_if_you_complete")), classes = Some("govuk-body-l"))
-            @button("global.label.continue_button", href = Some(routes.ApplicationController.uplift(Some(SafeRedirectUrl(redirectUrl))).url), id = "continue")
+            @button("global.label.continue_button", href = Some(routes.ApplicationController.uplift(Some(RedirectUrl(redirectUrl))).url), id = "continue")
```
Comment (vfryer12) [inline app/views/interstitial/ViewChildBenefitsSummarySingleAccountInterstitialView.scala.html:94]: The import needs to be updated to @import uk.gov.hmrc.play.bootstrap.binders.RedirectUrl

File: app/views/iv/start/ivStartPage.scala.html [hunk 3]
```diff
@@ -33,6 +33,6 @@
     @h1("label.confirm_your_identity")
     @p(Text(messages("label.before_you_continue_we_need_to_check")))
     @p(Text(messages("label.we_re_going_to_ask_you_some_security_questions")))
-    @button("global.label.continue_button", href = Some(controllers.routes.ApplicationController.uplift(redirectUrl.map(SafeRedirectUrl.apply)).url),  id="continue")
+    @button("global.label.continue_button", href = Some(controllers.routes.ApplicationController.uplift(redirectUrl.map(RedirectUrl.apply)).url),  id="continue")
```
Comment (vfryer12) [inline app/views/iv/start/ivStartPage.scala.html:36]: The import needs to be updated to @import uk.gov.hmrc.play.bootstrap.binders.RedirectUrl

File: app/views/interstitial/ViewNationalInsuranceInterstitialHomeView.scala.html [hunk 4]
```diff
@@ -67,6 +67,6 @@
         </div>
     } else {
         @p(Text(messages("label.you_can_see_this_part_of_your_account_if_you_complete")))
-        @button(href = Some(routes.ApplicationController.uplift(Some(SafeRedirectUrl(redirectUrl))).url), id="continue", msg = "global.label.continue_button")
+        @button(href = Some(routes.ApplicationController.uplift(Some(RedirectUrl(redirectUrl))).url), id="continue", msg = "global.label.continue_button")
```
Comment (vfryer12) [inline app/views/interstitial/ViewNationalInsuranceInterstitialHomeView.scala.html:70]: The import needs to be updated to @import uk.gov.hmrc.play.bootstrap.binders.RedirectUrl

**Review by vfryer12 — APPROVED** [review]

File: app/controllers/ApplicationController.scala [hunk 5]
```diff
@@ -48,63 +48,78 @@ class ApplicationController @Inject() (
...
-        case _         =>
-          logger.error("journeyId missing or incorect")
-          Future.successful(InternalServerError(technicalIssuesView(retryUrl)))
+      val safeUrl: Either[String, Option[SafeRedirectUrl]] = continueUrl match {
+        case None    => Right(None)
+        case Some(x) => x.getEither(OnlyRelative).map(Some.apply)
+      }
+
+      safeUrl match {
+        case Left(error)    => Future.successful(BadRequest(error))
```
Comment (grantrobson) [inline app/controllers/ApplicationController.scala:69]: This is quite complex. Cyclomatic complexity 17. I'd separate into private method

File: app/controllers/SelfAssessmentController.scala [hunk 6]
```diff
@@ -66,25 +67,33 @@ class SelfAssessmentController @Inject() (
         }
     }
 
-  def ivExemptLandingPage(continueUrl: Option[SafeRedirectUrl]): Action[AnyContent] =
+  def ivExemptLandingPage(continueUrl: Option[RedirectUrl]): Action[AnyContent] =
     authJourney.authWithPersonalDetails { implicit request =>
-      val retryUrl = routes.ApplicationController.uplift(continueUrl).url
+      val safeUrl: Either[String, Option[SafeRedirectUrl]] = continueUrl match {
```
Comment (grantrobson) [inline app/controllers/SelfAssessmentController.scala:72]: As per previous comment, can be simplified

**Review by grantrobson — APPROVED** [review]

**Pull request merged**